colexechash: optimize the hash aggregator and unordered distinct

yuzefovich opened the pull request:
**colexechash: some mechanical changes to the hash table**

This commit moves one method into the templated code and inlines another
method. This is a purely mechanical change in preparation for the
follow-up commit that will be templating the touched code.

Additionally, this commit simplifies the probing logic a bit by
squashing two `if` blocks into one via duplicating the code that needs to
run in some cases in both arms of the conditional.

Release note: None

**colexechash: optimize the hash aggregator and unordered distinct**

This commit optimizes the hash table code when it is used by the hash
aggregator or the unordered distinct. The following changes are made:
- now all tuples that are known to be distinct before probing (when the
tuple doesn't have a hash match) are omitted from probing entirely.
- the case when we're probing the batch against itself (in order to find
duplicates within the batch) has been optimized by taking advantage of
the following fact: if probing of `i`th tuple against `groupID[i]` tuple
revealed a mismatch, then we can skip probing against all tuples that
have non-zero `HeadID` - all such tuples are parts of the already
established equality chains, and it is guaranteed that the `i`th tuple
will be different from all of them too.
- a redundant `if` conditional is removed.

Additionally, in order to improve the chance of hash collisions in the
unit tests, this commit plumbs the testing knobs to reduce the number of
buckets used by the hash table.

Release note: None

yuzefovich committed:
colexechash: optimize the hash aggregator and unordered distinct

This commit optimizes the hash table code when it is used by the hash
aggregator or the unordered distinct. The following changes are made:
- now all tuples that are known to be distinct before probing (when the
tuple doesn't have a hash match) are omitted from probing entirely.
- the case when we're probing the batch against itself (in order to find
duplicates within the batch) has been optimized by taking advantage of
the following fact: if probing of `i`th tuple against `groupID[i]` tuple
revealed a mismatch, then we can skip probing against all tuples that
have non-zero `HeadID` - all such tuples are parts of the already
established equality chains, and it is guaranteed that the `i`th tuple
will be different from all of them too.
- a redundant `if` conditional is removed.

Additionally, in order to improve the chance of hash collisions in the
unit tests, this commit plumbs the testing knobs to reduce the number of
buckets used by the hash table.

Release note: None

yuzefovich committed:
colexechash: some mechanical changes to the hash table

This commit moves one method into the templated code and inlines another
method. This is a purely mechanical change in preparation for the
follow-up commit that will be templating the touched code.

Additionally, this commit simplifies the probing logic a bit by
squashing two `if` blocks into one via duplicating the code that needs to
run in some cases in both arms of the conditional.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69729)
<!-- Reviewable:end -->

yuzefovich commented:
The microbenchmarks are [here](https://gist.github.com/yuzefovich/cb8657c9f60c7db4c45ebf1d5b0360ff). I'm a bit surprised to not see much of a change for the hash aggregator - it's possible that our microbenchmarks don't capture the scenario where we would see the improvement. I plan to run some actual queries to see what's up.

yuzefovich committed:
WIP on more optimization

yuzefovich committed:
WIP on cleaning up the optimization

yuzefovich committed:
WIP on adding BCE

yuzefovich committed:
WIP adjust the benchmark

yuzefovich committed:
colexec: switch min to any_not_null for aggregator benchmarks

`any_not_null` will have the least impact when benchmarking the
aggregators.

Release note: None

yuzefovich committed:
WIP on reducing the number of buckets in hash agg (for benchmarking)

yuzefovich commented:
I figured out why the improvement on the unordered distinct benchmark is so good - it is because of the way we construct the input data set. Namely, for both ordered and unordered we construct the input in the same fashion, just for the latter we don't take the ordering property into account; however, this PR improves the "probing of the batch against itself" step, and since all data is ordered, we have a lot of duplicates within the batches.

I'll try truly randomizing the unordered distinct input data set and see what happens.

yuzefovich committed:
colexec: shuffle input for unordered distinct benchmark

Previously, we were using the same way to construct the input data set
for both ordered and unordered distinct benchmarks (the latter was just
not taking an explicit advantage of the ordering). However, this is not
very realistic - we should properly shuffle the input for the unordered
distinct which is what this commit does.

Release note: None

yuzefovich commented:
superseded by #101204
